lib/common/zstd_internal.h

```diff
@@ -209,33 +209,41 @@ typedef enum {
 /*! ZSTD_wildcopy() :
  *  custom version of memcpy(), can overwrite up to WILDCOPY_OVERLENGTH bytes (if length==0) */
 MEM_STATIC FORCE_INLINE_ATTR DONT_VECTORIZE
-void ZSTD_wildcopy(void* dst, const void* src, ptrdiff_t length, ZSTD_overlap_e ovtype)
+void ZSTD_wildcopy(void* dst, const void* src, BYTE* oend_g, ptrdiff_t length, ZSTD_overlap_e ovtype)
 {
     ptrdiff_t diff = (BYTE*)dst - (const BYTE*)src;
     const BYTE* ip = (const BYTE*)src;
     BYTE* op = (BYTE*)dst;
     BYTE* const oend = op + length;
 
     assert(diff >= 8 || (ovtype == ZSTD_no_overlap && diff < -8));
+
     if (length < VECLEN || (ovtype == ZSTD_overlap_src_before_dst && diff < VECLEN)) {
       do
           COPY8(op, ip)
       while (op < oend);
     }
     else {
-      if ((length & 8) == 0)
-        COPY8(op, ip);
-      do {
-        COPY16(op, ip);
+      if (oend < oend_g-16) {
```

Comment:
__builtin_expect(cond, 1)?

Reply:
That does not seem like a bad idea but no other code in zstd uses __builtin_expect except some code in pzstd that comes from the folly library.  I'm not opposed to it but it does not seem to fit zstd coding style.

Reply:
If it improves code generation, then we should use it. But I rarely see cases where it actually helps.

Reply:
https://godbolt.org/z/bP_kP0 in this case, Clang's assemblery looks nice. Or maybe https://godbolt.org/z/QEBBzG is better, hard to say.

If we take assert as "assumes" to help codegen, we see something like:
https://godbolt.org/z/QEBBzG

Hopefully next week I will dive into zstd's benchmarks :) But I think one of these codegens is better than current one :)

Reply:
@davidbolvansky IIRC @Cyan4973 has experimented with assuming asserts, but didn't find a meaningful perf impact. But that was a while ago, and things may have changed.

Reply:
Indeed, nowadays' branch predictors are extremely efficient, and it's generally a bad idea to try to outsmart them. I've seen `expect()` produce negative outcome more often than positive one.

Now, `expect()` also has an impact on code layout, and this one is baked in the binary: the cpu can't change that one, having the right path following the branch can have subtle consequences on instruction flow.

Even then, it's better to back a case with hard numbers, measurements, and when applicable, a comparison of generated assembly.

So, `expect()` is definitely not forbidden, just it has a high bar to meet.

Reply:
I took current zstd's master

gcc-9 and selesia

time ./zstd -d b/*.zst -f 

real	0m0,414s
user	0m0,273s
sys	0m0,141s

ASSUME version

xbolva00@xbolva00-G551JW:~/zstd$ time ./zstd -d b/*.zst -f 

real	0m0,407s
user	0m0,294s
sys	0m0,113s
xbolva00@xbolva00-G551JW:~/zstd$ time ./zstd -d b/*.zst -f 

real	0m0,405s
user	0m0,297s
sys	0m0,109s
xbolva00@xbolva00-G551JW:~/zstd$ time ./zstd -d b/*.zst -f 

real	0m0,401s
user	0m0,273s
sys	0m0,129s

Clang

Before:
time ./zstd -d b/*.zst -f 

real	0m0,448s
user	0m0,332s
sys	0m0,116s



After
 time ./zstd -d b/*.zst -f 

real	0m0,439s
user	0m0,323s
sys	0m0,116s

Reply:
That's a good start @davidbolvansky !
I would also suggest to use an in-memory benchmark, instead of timing the CLI.
At this level of difference, the overhead of process launching and I/O seems very large compared to the potential gains. The in-memory benchmark should help reduce the noise.

Reply:
Yeah, I will use it.

Decompr. speed with Clang 8
time ./zstd -d b/*.zst -f

real	0m0,476s
user	0m0,360s
sys	0m0,116s

Decompr. speed with Clang 9 (well, actually 10 svn today's build)
time ./zstd -d b/*.zst -f

real	0m0,438s
user	0m0,302s
sys	0m0,137s

Reply:
zstd's master

GCC 9 LTO
time ./zstd -d b/*.zst -f

real	0m0,409s
user	0m0,290s
sys	0m0,119s

GCC 9 LTO PGO
time ./zstd -d b/*.zst -f

real	0m0,398s
user	0m0,286s
sys	0m0,112s

So PGO gives +-3% oerf boost